homeassistant/components/knx/config_flow.py

```diff
@@ -100,6 +105,11 @@ async def async_step_connection_type(
     ) -> FlowResult:
         """Handle connection type configuration."""
         if user_input is not None:
+            self._async_scan_gen = None  # stop the scan
+            if self._gatewayscanner:
+                self._found_gateways = list(
+                    self._gatewayscanner.found_gateways.values()
+                )
```

Comment:
Why not use the generator you just created? Within some short timeout?

Reply:
We already burned the first result of the generator to stop waiting.

Reply:
Could just store the first output, then continue the iterator here with an async for. That way it gets closed properly too

Reply:
I'll give it a try.

It would async block again then until the timeout is reached, whereas in current implementation the user cancels the search when proceeding. 
Not sure if that's even noticeable with 2 seconds though, but not exactly what I wanted. 
🤔 If closed properly we could in the future even use a longer / no timeout and send multiple search requests (eg. every second) like ETS does.

Reply:
Did you try the iteratior approach? That way you dont need to store the scanner object at all.

You could either let the scan function have a time limit argument, or wrap it in a another async generator like.

with supress(Timeout), async_timeout(1):
  yield from scan()

Reply:
It has to be stored because I want to advance to the selection phase of that step just when the first gateway is found - and this has to be awaited before to be able to add "automatic" to the available options. 
The step method has to return for that - to be called again with `user_input` when the selection is done.

The current approach (before this PR) doesn't use a generator but just awaits the timeout. For users with only 1 gateway (most users probably) this results in unnecessary wait time.

Reply:
Yes, but i just meant only the iterator needs to be stored. Not the gateway object. But i didnt look closely.

Reply:
Ah I see, have misread that.
Tbh I do think the properly aclosed iterator solution is the nicest. That way, as you said, we also avoid hidden changes in the lib - like a change in the timeout or such. And we have 0 wait time after user action.